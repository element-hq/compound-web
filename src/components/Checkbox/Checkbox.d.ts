import React, { PropsWithChildren } from "react";
type CheckboxProps = {
    kind?: "primary" | "critical";
    className?: string;
} & React.ComponentPropsWithoutRef<"input">;
export declare const Checkbox: ({ kind, className, ...props }: PropsWithChildren<CheckboxProps>) => React.JSX.Element;
export {};
