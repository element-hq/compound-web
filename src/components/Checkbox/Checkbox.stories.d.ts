import React from "react";
declare const _default: import("@storybook/types").ComponentAnnotations<import("@storybook/react/dist/types-0a347bb9").R, {
    kind?: "critical" | "primary" | undefined;
    className?: string | undefined;
    onMouseDown?: ((e: React.MouseEvent<HTMLInputElement, MouseEvent>) => void) | undefined;
} & Omit<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, "ref"> & {
    children?: React.ReactNode;
}>;
export default _default;
export declare const Primary: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, {
    kind?: "critical" | "primary" | undefined;
    className?: string | undefined;
    onMouseDown?: ((e: React.MouseEvent<HTMLInputElement, MouseEvent>) => void) | undefined;
} & Omit<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, "ref"> & {
    children?: React.ReactNode;
}>;
export declare const Critical: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, {
    kind?: "critical" | "primary" | undefined;
    className?: string | undefined;
    onMouseDown?: ((e: React.MouseEvent<HTMLInputElement, MouseEvent>) => void) | undefined;
} & Omit<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, "ref"> & {
    children?: React.ReactNode;
}>;
//# sourceMappingURL=Checkbox.stories.d.ts.map