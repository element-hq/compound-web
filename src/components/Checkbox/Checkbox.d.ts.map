{"version":3,"file":"Checkbox.d.ts","sourceRoot":"","sources":["../../../../src/components/Checkbox/Checkbox.tsx"],"names":[],"mappings":"AAiBA,OAAO,KAAK,EAAE,EAAE,iBAAiB,EAAE,MAAM,OAAO,CAAC;AAIjD,KAAK,aAAa,GAAG;IACnB,IAAI,CAAC,EAAE,SAAS,GAAG,UAAU,CAAC;IAC9B,SAAS,CAAC,EAAE,MAAM,CAAC;IACnB,WAAW,CAAC,EAAE,CAAC,CAAC,EAAE,KAAK,CAAC,UAAU,CAAC,gBAAgB,EAAE,UAAU,CAAC,KAAK,IAAI,CAAC;CAC3E,GAAG,KAAK,CAAC,wBAAwB,CAAC,OAAO,CAAC,CAAC;AAE5C,eAAO,MAAM,QAAQ,+CAKlB,kBAAkB,aAAa,CAAC,sBAmBlC,CAAC"}