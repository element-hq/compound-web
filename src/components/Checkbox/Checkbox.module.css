/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.checkbox {
  --size: 24px;

  display: grid;
  width: var(--size);
  height: var(--size);
}

.checkbox > * {
  grid-area: 1/1;
  width: var(--size);
  height: var(--size);
}

.checkbox > input[type="checkbox"] {
  opacity: 0;
  margin: 0;
  cursor: pointer;
}

.checkbox-ui {
  pointer-events: none;
  width: var(--size);
  box-sizing: border-box;
  aspect-ratio: 1 / 1;
  border-radius: 4px; /* TODO: Ought to be a token */
  border: 1px solid;
}

:checked + .checkbox-ui {
  color: inherit; /* TODO: Replace with the checkmark code */
}

[data-kind="primary"] .checkbox-ui {
  border-color: var(--cpd-color-border-interactive-primary);
}

[data-kind="critical"] .checkbox-ui {
  border-color: var(--cpd-color-border-critical-primary);
}

[type="checkbox"]:focus + .checkbox-ui {
  outline: 2px solid var(--cpd-color-blue-900);
  outline-offset: 1px;
}

.checkbox-ui svg {
  width: var(--size);
  height: var(--size);
}

.checkbox :not(:checked) + .checkbox-ui {
  color: transparent;
}

.checkbox :checked + .checkbox-ui {
  color: var(--cpd-color-icon-on-solid-primary);
}

.checkbox[data-kind="primary"] :checked + .checkbox-ui {
  background-color: var(--cpd-color-bg-action-primary-rest);
  border-color: var(--cpd-color-bg-action-primary-rest);
}

.checkbox[data-kind="critical"] :checked + .checkbox-ui {
  background-color: var(--cpd-color-bg-critical-primary);
  border-color: var(--cpd-color-bg-critical-primary);
}

@media (hover) {
  .checkbox[data-kind="primary"] [type="checkbox"]:not[disabled]:hover + .checkbox-ui {
    color: var(--cpd-color-icon-quaternary);
    border-color: var(--cpd-color-border-interactive-hovered);
  }
}

.checkbox[data-kind="critical"] [type="checkbox"]:active + .checkbox-ui {
  color: var(--cpd-color-icon-quaternary);
  border-color: var(--cpd-color-border-critical-hovered);
}

.checkbox[data-kind] [type="checkbox"][disabled] + .checkbox-ui {
  border-color: var(--cpd-color-border-disabled);
  background: var(--cpd-color-bg-canvas-disabled);
}

.checkbox[data-kind="primary"] [type="checkbox"]:not[disabled]:active + .checkbox-ui {
  color: var(--cpd-color-icon-quaternary);
  border-color: var(--cpd-color-border-interactive-hovered);
}

@media (hover) {
  .checkbox[data-kind="critical"] [type="checkbox"]:not[disabled]:hover + .checkbox-ui {
    color: var(--cpd-color-icon-quaternary);
    border-color: var(--cpd-color-border-critical-hovered);
  }
}

.checkbox[data-kind] [type="checkbox"][disabled]:checked + .checkbox-ui {
  background: var(--cpd-color-bg-action-primary-disabled);
  color: var(--cpd-color-text-disabled);
}

@media (hover) {
  .checkbox[data-kind="primary"]
    [type="checkbox"]:not[disabled]:checked:hover
    + .checkbox-ui {
    background: var(--cpd-color-bg-action-primary-hovered);
  }
}

.checkbox[data-kind="primary"] [type="checkbox"]:not[disabled]:checked:active + .checkbox-ui {
  background: var(--cpd-color-bg-action-primary-hovered);
}

@media (hover) {
  .checkbox[data-kind="critical"]
    [type="checkbox"]:not[disabled]:checked:hover
    + .checkbox-ui {
    background: var(--cpd-color-bg-critical-hovered);
  }
}

.checkbox[data-kind="critical"]
  [type="checkbox"]:not[disabled]:checked:active
  + .checkbox-ui {
  background: var(--cpd-color-bg-critical-hovered);
}