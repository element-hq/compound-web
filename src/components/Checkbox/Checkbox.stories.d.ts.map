{"version":3,"file":"Checkbox.stories.d.ts","sourceRoot":"","sources":["../../../../src/components/Checkbox/Checkbox.stories.tsx"],"names":[],"mappings":"AAgBA,OAAO,KAAK,MAAM,OAAO,CAAC;;;;;;;;AAK1B,wBAKoC;AAMpC,eAAO,MAAM,OAAO;;;;;;EAAoB,CAAC;AAIzC,eAAO,MAAM,QAAQ;;;;;;EAAoB,CAAC"}