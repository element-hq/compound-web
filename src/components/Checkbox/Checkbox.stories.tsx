/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { Meta, StoryFn } from "@storybook/react";

import { Checkbox as CheckboxComponent } from "./Checkbox";

export default {
  title: "Checkbox",
  component: CheckboxComponent,
  argTypes: {},
  args: {},
} as Meta<typeof CheckboxComponent>;

const Template: StoryFn<typeof CheckboxComponent> = (args) => (
  <CheckboxComponent {...args} />
);

export const Primary = Template.bind({});
Primary.args = {};
Primary.parameters = {};

export const Critical = Template.bind({});
Critical.args = {
  kind: "critical",
};
Critical.parameters = {};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
  checked: false,
};

export const DisabledChecked = Template.bind({});
DisabledChecked.args = {
  disabled: true,
  checked: true,
};
