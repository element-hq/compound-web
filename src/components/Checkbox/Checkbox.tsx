/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import React, { PropsWithChildren } from "react";
import styles from "./Checkbox.module.css";
import CheckIcon from "@vector-im/compound-design-tokens/icons/check.svg";

type CheckboxProps = {
  kind?: "primary" | "critical";
  className?: string;
  onMouseDown?: (e: React.MouseEvent<HTMLInputElement, MouseEvent>) => void;
} & React.ComponentPropsWithoutRef<"input">;

export const Checkbox: React.FC<PropsWithChildren<CheckboxProps>> = ({
  kind = "primary",
  className,
  onMouseDown,
  ...props
}) => {
  const classes = classnames(styles.checkbox, className);
  return (
    <div className={classes} data-kind={kind}>
      <input
        {...props}
        type="checkbox"
        onMouseDown={(e) => {
          // Prevents the `focus` event from being fired when clicked
          // but still reliably works when navigating via the keyboard
          e.preventDefault();
          onMouseDown?.(e);
        }}
      />
      <div className={styles["checkbox-ui"]}>
        <CheckIcon aria-hidden={true} className={styles["checkbox-check"]} />
      </div>
    </div>
  );
};
