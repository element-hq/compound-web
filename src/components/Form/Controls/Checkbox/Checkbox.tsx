/*
Copyright 2025 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import classnames from "classnames";
import React, { ComponentProps, ComponentRef, forwardRef } from "react";
import CheckIcon from "@vector-im/compound-design-tokens/assets/web/icons/check";
import { Control } from "@radix-ui/react-form";

import styles from "./Checkbox.module.css";

type Props = {
  /**
   * The CSS class name.
   */
  className?: string;
} & Omit<ComponentProps<"input">, "type">;

/**
 * A styled checkbox input, for standalone use.
 */
export const CheckboxInput = forwardRef<HTMLInputElement, Props>(
  function Checkbox({ className, ...props }, ref) {
    const classes = classnames(styles.container, className);
    return (
      <div className={classes}>
        <input ref={ref} className={styles.input} {...props} type="checkbox" />
        <div className={styles.ui}>
          <CheckIcon aria-hidden={true} />
        </div>
      </div>
    );
  },
);

/**
 * A styled checkbox input wrapped in a `Control` component, for use in Radix forms.
 */
export const CheckboxControl = forwardRef<
  ComponentRef<typeof CheckboxInput>,
  ComponentProps<typeof CheckboxInput>
>(function CheckboxControl(props, ref) {
  return (
    <Control asChild>
      <CheckboxInput ref={ref} {...props} />
    </Control>
  );
});
