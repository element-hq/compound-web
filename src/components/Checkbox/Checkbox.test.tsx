/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { getByRole, render } from "@testing-library/react";
import React from "react";

import { Checkbox } from "./Checkbox";
import userEvent from "@testing-library/user-event";

describe("Checkbox", () => {
  it("renders", () => {
    const { asFragment } = render(<Checkbox />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not have focus after being clicked", async () => {
    const { container } = render(<Checkbox />);
    const checkbox = getByRole(container, "checkbox");
    await userEvent.click(checkbox);
    expect(checkbox).not.toHaveFocus();
  });
});
