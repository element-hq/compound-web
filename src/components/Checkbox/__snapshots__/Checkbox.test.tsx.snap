// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders 1`] = `
<DocumentFragment>
  <div
    class="checkbox"
    data-kind="primary"
  >
    <input
      type="checkbox"
    />
    <div
      class="checkbox-ui"
    >
      <svg
        aria-hidden="true"
        class="checkbox-check"
      />
    </div>
  </div>
</DocumentFragment>
`;
